Upgrade to recent FluentD and Ruby versions

markbreedlove opened the pull request:
This PR makes a number of changes so that the formula works with more recent o/s, Ruby, and FluentD versions. It also makes sure that FluentD runs as the configured user. (It was creating the user, but not ensuring that the daemon runs as that user.)

Updates:

* Use `rbenv` to install the same currently-updated version of Ruby, regardless of o/s.
* Run `fluentd` as the configured user, by default `fluentd`.
* Update Vagrant boxes to validate the formula on modern Debian, Ubuntu, and CentOS machines.
* Fix the syntax of FluentD config files to account for changes from version 0.x to version 1.x.
* Fix Salt states to account for changes in rendering behavior in either Salt or Jinja.
* Fix issues with service starts and restarts.

You should be able to bring up all three Vagrant VMs with `vagrant up` and the `testinfra` suites should pass.

Commit:
Disable Guest Additions auto-update

Commit:
Remove invalid -Z option from salt bootstrap

Commit:
Update Vagrant boxes

Use the officially-recommended bento boxes, per
https://www.vagrantup.com/docs/boxes.html#official-boxes

Commit:
Add tojson filter, fix unicode string repr

Commit:
Fix Salt installation

* Add /etc/hosts entry for "salt"
* Specify git installation of specific good version (most recent has bug)

Commit:
Add more tojson filters to fix string reprs

Commit:
Back out addition of unnecessary tojson filter

Commit:
Make compatible w modern FluentD

Update Salt pillar, testinfra script, and tests so that they work with
current versions of FluentD.

Commit:
Add __pycache__ exclusion to .gitignore

Commit:
Fix service restart issues

Commit:
Fix startup and privilege separation, use supported Ruby

* Make FluentD run as configured user, under rbenv and modern Ruby version
* Turn service control files into templates to add user parameter
* Change CentOS VM back to version 7
* Fix service starts and restarts by tuning timeouts
* Use recommended /run directory for pid files instead of /var/run

blarghmatey commented on Vagrantfile:
```diff
@@ -82,11 +95,13 @@ Vagrant.configure(2) do |config|
   config.vm.provision "shell", path: "scripts/vagrant_setup.sh"
   config.vm.provision :salt do |salt|
-    salt.bootstrap_options = '-U -Z'
+    salt.bootstrap_options = '-U'
```
We should be adding `-x python3` to ensure that we're testing against a Python 3 version of Salt.

markbreedlove replied:
Making the 1-line change to use Python 3 produces this error when trying to build a new VM:

```
 *  INFO: Running config_salt()
 *  INFO: Running install_debian_git()
Traceback (most recent call last):
  File "setup.py", line 26, in <module>
    import distutils.dist
ModuleNotFoundError: No module named 'distutils.dist'
 * ERROR: Failed to run install_debian_git()!!!
```

Does this ring a bell at all?
`distutils.dist` is part of the standard library, so I'm not sure what to think yet.

markbreedlove replied:
I made it use Salt's package by removing `salt.install_type` and `salt.install_args` (see below), but now I get this error:

```
The following packages have unmet dependencies:
 salt-minion : Depends: salt-common (= 2019.2.0+ds-1) but it is not going to be installed
 * ERROR: Failed to run install_debian_stable()!!!
```

... pondering ...

markbreedlove replied:
It's a known issue with Debian 10. I'll downgrade to Debian 9.

markbreedlove replied:
I've added `-x python3`.

blarghmatey commented on Vagrantfile:
```diff
@@ -82,11 +95,13 @@ Vagrant.configure(2) do |config|
     salt.colorize = true
     salt.verbose = true
+    salt.install_type = 'git'
+    salt.install_args = 'v2019.2.0'
```
We can safely remove this as they have reverted the `latest` package to be 2019.2.0

blarghmatey commented on fluentd/map.jinja:
```diff
@@ -11,24 +11,65 @@
         },
         'ssl_directory': '/etc/ssl',
-        'nginx_site_path': '/etc/nginx/sites-enabled'
+        'nginx_site_path': '/etc/nginx/sites-enabled',
```
We should excise the Nginx functionality from this formula in favor of composing it with the Nginx formula so that we aren't inflating the functionality needlessly.

blarghmatey commented on fluentd/map.jinja:
```diff
@@ -11,24 +11,65 @@
         ]
     },
-}, grain='os_family', merge=salt.pillar.get('fluentd:overrides'), base='default') %}
+}, grain='os', merge=salt.pillar.get('fluentd:overrides'), base='default') %}
```
Why the change to `os` from `os_family`?

markbreedlove replied:
This was due to there being differences in the names of some development packages between Ubuntu and Debian, but they're not needed anymore, following the decision not to use Rbenv.

blarghmatey commented on scripts/vagrant_setup.sh:
```diff
@@ -35,3 +35,6 @@ base:
     - fluentd.reverse_proxy" | sudo tee /srv/salt/top.sls
+
+# The salt-minion service needs this:
+sudo sh -c 'echo "127.0.0.1 salt" >> /etc/hosts'
```
You can remove this in favor of adding `-A 127.0.0.1` to the script args in the vagrantfile

blarghmatey commented on tests/conftest.py:
```diff
@@ -3,16 +3,16 @@
 
 @pytest.fixture()
-def RubyGem(Command):
+def RubyGem(host):
```
We should be able to remove this entire file in favor of the rubygem resource built into Testinfra in recent releases.

markbreedlove replied:
I'm having a hard time finding any documentation on that. Can you point me in the right direction?

blarghmatey replied:
I was mistaken. There had been some back and forth a while back and I thought that a gem module had been added. We can take a look at upstreaming this later.

blarghmatey reviewed (changes requested):
For the testinfra tests, we should translate them to use the integration with Salt. The pytest style tests predate that extension and all of the more recent formulas use the salt approach.

Commit:
Don't use git installation of Salt

The problematic Salt release has been withdrawn, so we can use the
"latest" one from their package, instead of installing with git.

A side effect of this is that we have to drop down to Debian 9 for the
"debian" box. As we do that, we have to change the package name of
libncurses-dev to libncurses5-dev.

Commit:
Make compatible with Python 3

Update the FluentD config file Jinja2 template so that it works with
Python 3 as well as Python 2, and update the list of o/s package
dependencies.

Update the Vagrantfile to have Salt run under Python 3.

The changes to the Jinja2 template are to adapt to changes in the
OrderedDict class with Python 3. The template still works with Python 2.

Commit:
Do not use Rbenv

Take out Rbenv, which was added in this PR. Use the system Ruby.

The only problem with this is that we have to comment out the CentOS VM,
because getting fluentd to work there is just too problematic. There is
no Salt package for CentOS 8, and CentOS 7 has too old a version of Ruby
(2.0.0) for fluentd itself and a number of its plugins. I was not able
to get fluentd to start after backing way down to the 0.12 release and
disabling all plugins. Our time will be better spent waiting for
SaltStack to release a package for CentOS 8.

Commit:
Change salt master resolution in Vagrant boxes

Add a parameter to the Salt bootstrap script in the Vagrantfile to
identify the master's IP address, rather than amending /etc/hosts in
vagrant_setup.sh.

blarghmatey commented:
The only thing left here is to convert the tests to use the testinfra plugin, but we can handle that as a separate issue.

blarghmatey commented:
:+1:

markbreedlove commented:
This is ready to merge and release, but I'm holding off until https://github.com/mitodl/salt-ops/pull/539 looks like it's mostly ready to go, in order to coordinate their release & deployment.

blarghmatey commented:
@markbreedlove this looks ready to merge and shouldn't have any impact on existing infrastructure so there's no need to wait for the salt-ops/pillar piece of things. I'm actually running into a bug due to Salt on Python3 that this will fix so I'm merging now. Thanks!
